sonic-net/sonic-utilities
# Several fixes and updates for ecnconfig

#### What I did
- Update DB entry table ref syntax.
- Allow non-lossless queues to be queried and updated.
- Remove entire wred_profile entry to activate a change to the SAI layer.
- If wred_profile was the only value and was removed, delete entire key entry.
- Check that a port is available before attempting an index access.

#### How I did it

#### How to verify it
- Update UT to reflect new expectations and DB syntax.
- Add new tests for lossy queues.
- Manually execute a variety of commands to observe output. 

#### Previous command output (if the output of a command-line utility has changed)

#### New command output (if the output of a command-line utility has changed)
ecnconfig -q 3 off

syslog:
```
Jun 29 16:25:43.287115 sonic NOTICE swss#orchagent: :- handleQueueTable: QUEUE|Ethernet60|3 wred_profile was configured but is not any more. Remove it
Jun 29 16:25:43.294859 sonic NOTICE swss#orchagent: :- handleQueueTable: QUEUE|Ethernet76|3 wred_profile was configured but is not any more. Remove it
```
sairedis.rec:
```
2023-06-29.16:25:43.284060|s|SAI_OBJECT_TYPE_SCHEDULER_GROUP:oid:0x170000000002c9|SAI_SCHEDULER_GROUP_ATTR_SCHEDULER_PROFILE_ID=oid:0x160000000004df
2023-06-29.16:25:43.286168|s|SAI_OBJECT_TYPE_QUEUE:oid:0x15000000000128|SAI_QUEUE_ATTR_WRED_PROFILE_ID=oid:0x130000000004e2
2023-06-29.16:25:43.287026|s|SAI_OBJECT_TYPE_SCHEDULER_GROUP:oid:0x170000000002c9|SAI_SCHEDULER_GROUP_ATTR_SCHEDULER_PROFILE_ID=oid:0x160000000004df
2023-06-29.16:25:43.287794|s|SAI_OBJECT_TYPE_QUEUE:oid:0x15000000000128|SAI_QUEUE_ATTR_WRED_PROFILE_ID=oid:0x0
2023-06-29.16:25:43.294791|s|SAI_OBJECT_TYPE_SCHEDULER_GROUP:oid:0x170000000002ed|SAI_SCHEDULER_GROUP_ATTR_SCHEDULER_PROFILE_ID=oid:0x160000000004df
2023-06-29.16:25:43.295568|s|SAI_OBJECT_TYPE_QUEUE:oid:0x15000000000168|SAI_QUEUE_ATTR_WRED_PROFILE_ID=oid:0x0
```

ecnconfig -q 3 on

sairedis.rec:
```
2023-06-29.16:27:57.892976|s|SAI_OBJECT_TYPE_SCHEDULER_GROUP:oid:0x170000000002c9|SAI_SCHEDULER_GROUP_ATTR_SCHEDULER_PROFILE_ID=oid:0x160000000004df
2023-06-29.16:27:57.894043|s|SAI_OBJECT_TYPE_QUEUE:oid:0x15000000000128|SAI_QUEUE_ATTR_WRED_PROFILE_ID=oid:0x130000000004e2
2023-06-29.16:27:57.901807|s|SAI_OBJECT_TYPE_SCHEDULER_GROUP:oid:0x170000000002ed|SAI_SCHEDULER_GROUP_ATTR_SCHEDULER_PROFILE_ID=oid:0x160000000004df
2023-06-29.16:27:57.902492|s|SAI_OBJECT_TYPE_QUEUE:oid:0x15000000000168|SAI_QUEUE_ATTR_WRED_PROFILE_ID=oid:0x130000000004e2
```

Updating non-lossless TCs, specifically a TC that doesn't have a queue entry and ensuring it returns to its original state when turned off:
```
# redis-dump -d 4 -y -k "QUEUE|Ethernet60|7"
{}# ./ecnconfig -q 7 on
# redis-dump -d 4 -y -k "QUEUE|Ethernet60|7"
{
  "QUEUE|Ethernet60|7": {
    "expireat": 1688056336.0276697,
    "ttl": -0.001,
    "type": "hash",
    "value": {
      "wred_profile": "AZURE_LOSSLESS"
    }
  }
}# ./ecnconfig -q 7 off
# redis-dump -d 4 -y -k "QUEUE|Ethernet60|7"
{}# ./ecnconfig -q 7
ECN status:
Executing queue: 7
queue 7: off
#
```

Generic enabled profile confirming 'on' is reported:
```
# ./ecnconfig -q 3
ECN status:
queue 3: on
```

Lossless TC 3 toggling with DB output:
```
# ./ecnconfig -q 3
ECN status:
queue 3: on
# redis-dump -d 4 -y -k "QUEUE|Ethernet60|3"
{
  "QUEUE|Ethernet60|3": {
    "expireat": 1688062188.101967,
    "ttl": -0.001,
    "type": "hash",
    "value": {
      "scheduler": "scheduler.1",
      "wred_profile": "AZURE_LOSSLESS"
    }
  }
# ./ecnconfig -q 3 off
# redis-dump -d 4 -y -k "QUEUE|Ethernet60|3"
{
  "QUEUE|Ethernet60|3": {
    "expireat": 1688062199.867536,
    "ttl": -0.001,
    "type": "hash",
    "value": {
      "scheduler": "scheduler.1"
    }
  }
# ./ecnconfig -q 3 on
# redis-dump -d 4 -y -k "QUEUE|Ethernet60|3"
{
  "QUEUE|Ethernet60|3": {
    "expireat": 1688062221.2048986,
    "ttl": -0.001,
    "type": "hash",
    "value": {
      "scheduler": "scheduler.1",
      "wred_profile": "AZURE_LOSSLESS"
    }
  }
```
